hermes-frontend/src/test/groovy/pl/allegro/tech/hermes/frontend/publishing/DynamicThroughputLimiterTest.groovy

```diff
@@ -0,0 +1,119 @@
+package pl.allegro.tech.hermes.frontend.publishing
+
+import com.codahale.metrics.Metered
+import pl.allegro.tech.hermes.api.TopicName
+import pl.allegro.tech.hermes.frontend.publishing.handlers.DynamicThroughputLimiter
+import pl.allegro.tech.hermes.frontend.publishing.handlers.ThroughputLimiter
+import spock.lang.Specification
+
+import java.util.concurrent.ScheduledExecutorService
+
+class DynamicThroughputLimiterTest extends Specification {
+
+    def max = 10_000
+    def threshold = 8000
+    def desired = 6000
+    def idleThreshold = 0.5
+
+    double globalRate = 0.0
+    def globalMeter = [getOneMinuteRate: {globalRate}] as Metered
+
+    def topicName = new TopicName("group", "name")
+    def executor = Mock(ScheduledExecutorService)
+
+    def limiter = new DynamicThroughputLimiter(max, threshold, desired, idleThreshold, 1, globalMeter, executor)
+
+
+    def "global allow"() {
+        given:
+        globalRate = 10
+
+        expect:
+        limiter.check(topicName, meter(1000))
+    }
+
+    def "global block"() {
+        given:
+        globalRate = 12_000
```

Comment:
mocking using spock makes this operation very verbose because I can't dynamically change this value until test is over which violates idea of throughput that changes in time